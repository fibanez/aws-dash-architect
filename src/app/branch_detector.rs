//! Dynamic Branch Detection System
//!
//! This module provides comprehensive branch detection capabilities for git repositories,
//! designed to replace hardcoded branch assumptions with dynamic detection of the
//! default branch and fallback mechanisms for common branch naming conventions.

use anyhow::{anyhow, Result};
use git2::{Repository, Direction};
use tracing::{error, info, warn};

use crate::app::git_error_handling::{
    GitOperation, GitOperationError, GitOperationLogger
};

/// Information about a git branch
#[derive(Debug, Clone, PartialEq)]
pub struct BranchInfo {
    /// Branch name (without refs/heads/ prefix)
    pub name: String,
    /// Whether this is the default branch
    pub is_default: bool,
    /// Commit hash this branch points to
    pub commit_hash: String,
    /// Whether this is a remote branch
    pub is_remote: bool,
}

/// Dynamic branch detector for git repositories
pub struct BranchDetector {
    /// Repository URL to query
    repo_url: String,
    /// Common branch names to try as fallbacks
    fallback_branches: Vec<String>,
}

impl BranchDetector {
    /// Create a new branch detector for the given repository URL
    pub fn new(repo_url: String) -> Self {
        Self {
            repo_url,
            fallback_branches: vec![
                "main".to_string(),
                "master".to_string(),
                "develop".to_string(),
                "dev".to_string(),
                "trunk".to_string(),
            ],
        }
    }

    /// Create a new branch detector with custom fallback branches
    pub fn with_fallbacks(repo_url: String, fallback_branches: Vec<String>) -> Self {
        Self {
            repo_url,
            fallback_branches,
        }
    }

    /// Detect the default branch for the repository
    pub fn detect_default_branch(&self) -> Result<String> {
        // First, try to detect the default branch using remote HEAD reference
        match self.detect_default_branch_from_remote() {
            Ok(branch_name) => {
                info!("Successfully detected default branch: {}", branch_name);
                return Ok(branch_name);
            }
            Err(e) => {
                warn!("Failed to detect default branch from remote HEAD: {}", e);
            }
        }

        // If remote HEAD detection fails, try fallback branches
        match self.detect_branch_from_fallbacks() {
            Ok(branch_name) => {
                info!("Successfully detected branch using fallback: {}", branch_name);
                Ok(branch_name)
            }
            Err(e) => {
                error!("Failed to detect any valid branch: {}", e);
                Err(anyhow!("Failed to detect any valid branch: {}", e))
            }
        }
    }

    /// Get all available branches from the remote repository
    pub fn get_available_branches(&self) -> Result<Vec<BranchInfo>> {
        let mut logger = GitOperationLogger::start(
            GitOperation::BranchDetection,
            std::env::temp_dir(),
            Some(self.repo_url.clone()),
            None,
        );

        logger.add_context("operation".to_string(), "list_remote_branches".to_string());
        info!("Listing available branches for repository: {}", self.repo_url);

        // Create a temporary repository to list remote branches
        let temp_dir = tempfile::tempdir()
            .map_err(|e| anyhow!("Failed to create temporary directory: {}", e))?;
        
        let temp_repo_path = temp_dir.path().join("temp_repo");
        logger.add_context("temp_repo_path".to_string(), temp_repo_path.display().to_string());

        // Initialize a bare repository for remote operations
        let repo = Repository::init_bare(&temp_repo_path)
            .map_err(|e| {
                let git_error = GitOperationError::from_git2_error(
                    GitOperation::RepositoryOpen,
                    e,
                    temp_repo_path.clone(),
                    Some(self.repo_url.clone()),
                    None,
                );
                git_error.log_error();
                git_error.into_anyhow()
            })?;

        // Add remote
        let mut remote = repo.remote("origin", &self.repo_url)
            .map_err(|e| {
                let git_error = GitOperationError::from_git2_error(
                    GitOperation::RemoteOperation,
                    e,
                    temp_repo_path.clone(),
                    Some(self.repo_url.clone()),
                    None,
                );
                git_error.log_error();
                git_error.into_anyhow()
            })?;

        // Set up callbacks for listing references
        let mut branches = Vec::new();
        let mut default_branch: Option<String> = None;

        {
            let branches_ref = &mut branches;
            let default_branch_ref = &mut default_branch;
            
            // Use TLS-configured callbacks from git_error_handling module
            use crate::app::git_error_handling::tls_config;
            let mut callbacks = tls_config::configure_tls_callbacks();
            
            // Add branch detection logic on top of TLS configuration
            callbacks.update_tips(|refname, _old_oid, new_oid| {
                // Parse branch references
                if let Some(branch_name) = refname.strip_prefix("refs/heads/") {
                    branches_ref.push(BranchInfo {
                        name: branch_name.to_string(),
                        is_default: false, // Will be updated later
                        commit_hash: new_oid.to_string(),
                        is_remote: true,
                    });
                }
                
                true
            });

            // Connect to remote and list references
            remote.connect_auth(Direction::Fetch, Some(callbacks), None)
                .map_err(|e| {
                    let git_error = GitOperationError::from_git2_error(
                        GitOperation::RemoteOperation,
                        e,
                        temp_repo_path.clone(),
                        Some(self.repo_url.clone()),
                        None,
                    );
                    git_error.log_error();
                    git_error.into_anyhow()
                })?;

            // List remote references
            let refs = remote.list()
                .map_err(|e| {
                    let git_error = GitOperationError::from_git2_error(
                        GitOperation::BranchDetection,
                        e,
                        temp_repo_path.clone(),
                        Some(self.repo_url.clone()),
                        None,
                    );
                    git_error.log_error();
                    git_error.into_anyhow()
                })?;

            // Process references to find branches and default branch
            for remote_head in refs {
                let refname = remote_head.name();
                let oid = remote_head.oid();
                
                if let Some(branch_name) = refname.strip_prefix("refs/heads/") {
                    // Check if this branch already exists in our list
                    if let Some(existing_branch) = branches_ref.iter_mut().find(|b| b.name == branch_name) {
                        existing_branch.commit_hash = oid.to_string();
                    } else {
                        branches_ref.push(BranchInfo {
                            name: branch_name.to_string(),
                            is_default: false,
                            commit_hash: oid.to_string(),
                            is_remote: true,
                        });
                    }
                }
                
                // Check if this is the symbolic HEAD reference
                if refname == "HEAD" {
                    // Try to find which branch HEAD points to
                    for branch in branches_ref.iter_mut() {
                        if branch.commit_hash == oid.to_string() {
                            branch.is_default = true;
                            *default_branch_ref = Some(branch.name.clone());
                            break;
                        }
                    }
                }
            }

            let _ = remote.disconnect();
        }

        // If we couldn't determine the default branch from HEAD, use heuristics
        if default_branch.is_none() && !branches.is_empty() {
            // Try to find a branch matching our fallback list
            for fallback in &self.fallback_branches {
                if let Some(branch) = branches.iter_mut().find(|b| b.name == *fallback) {
                    branch.is_default = true;
                    default_branch = Some(branch.name.clone());
                    info!("Using fallback branch as default: {}", branch.name);
                    break;
                }
            }
            
            // If still no default, mark the first branch as default
            if default_branch.is_none() && !branches.is_empty() {
                branches[0].is_default = true;
                default_branch = Some(branches[0].name.clone());
            }
        }

        if branches.is_empty() {
            return Err(anyhow!("No branches found in remote repository"));
        }

        Ok(branches)
    }

    /// Validate that a specific branch exists in the repository
    pub fn validate_branch_exists(&self, branch_name: &str) -> Result<bool> {
        match self.get_available_branches() {
            Ok(branches) => {
                let exists = branches.iter().any(|b| b.name == branch_name);
                if !exists {
                    warn!("Branch '{}' does not exist in repository", branch_name);
                }
                Ok(exists)
            }
            Err(e) => {
                error!("Failed to validate branch existence: {}", e);
                Err(e)
            }
        }
    }

    /// Detect default branch using remote HEAD reference
    fn detect_default_branch_from_remote(&self) -> Result<String> {
        // Get all branches and find the default one
        let branches = self.get_available_branches()?;
        
        // Look for the branch marked as default
        if let Some(default_branch) = branches.iter().find(|b| b.is_default) {
            return Ok(default_branch.name.clone());
        }
        
        Err(anyhow!("Could not determine default branch from remote HEAD"))
    }

    /// Detect branch using fallback branch names
    fn detect_branch_from_fallbacks(&self) -> Result<String> {
        // Get all available branches
        let branches = self.get_available_branches()?;
        let branch_names: Vec<String> = branches.iter().map(|b| b.name.clone()).collect();
        
        // Try each fallback branch in order
        for fallback in &self.fallback_branches {
            if branch_names.contains(fallback) {
                return Ok(fallback.clone());
            }
        }
        
        // If no fallback matches, return the first available branch
        if let Some(first_branch) = branches.first() {
            return Ok(first_branch.name.clone());
        }
        
        Err(anyhow!("No valid branches found in repository"))
    }
}

/// Utility functions for branch detection
pub mod utils {

    /// Check if a branch name is likely to be a default branch
    pub fn is_likely_default_branch(branch_name: &str) -> bool {
        matches!(branch_name, "main" | "master" | "trunk" | "default")
    }

    /// Get a list of common default branch names
    pub fn get_common_default_branches() -> Vec<String> {
        vec![
            "main".to_string(),
            "master".to_string(),
            "trunk".to_string(),
            "default".to_string(),
        ]
    }

    /// Get a list of common development branch names
    pub fn get_common_dev_branches() -> Vec<String> {
        vec![
            "develop".to_string(),
            "dev".to_string(),
            "development".to_string(),
            "devel".to_string(),
        ]
    }

    /// Create a comprehensive list of fallback branches
    pub fn get_comprehensive_fallback_branches() -> Vec<String> {
        let mut branches = get_common_default_branches();
        branches.extend(get_common_dev_branches());
        branches
    }
}

#[cfg(test)]
mod tests {

    use super::{BranchDetector, BranchInfo, utils};

    #[test]
    fn test_branch_detector_creation() {
        let detector = BranchDetector::new("https://github.com/test/repo.git".to_string());
        assert_eq!(detector.repo_url, "https://github.com/test/repo.git");
        assert!(!detector.fallback_branches.is_empty());
        assert!(detector.fallback_branches.contains(&"main".to_string()));
        assert!(detector.fallback_branches.contains(&"master".to_string()));
    }

    #[test]
    fn test_branch_detector_with_custom_fallbacks() {
        let custom_fallbacks = vec!["custom".to_string(), "branch".to_string()];
        let detector = BranchDetector::with_fallbacks(
            "https://github.com/test/repo.git".to_string(),
            custom_fallbacks.clone()
        );
        assert_eq!(detector.fallback_branches, custom_fallbacks);
    }

    #[test]
    fn test_branch_info_creation() {
        let branch = BranchInfo {
            name: "main".to_string(),
            is_default: true,
            commit_hash: "abc123".to_string(),
            is_remote: true,
        };
        
        assert_eq!(branch.name, "main");
        assert!(branch.is_default);
        assert_eq!(branch.commit_hash, "abc123");
        assert!(branch.is_remote);
    }

    #[test]
    fn test_utils_functions() {
        assert!(utils::is_likely_default_branch("main"));
        assert!(utils::is_likely_default_branch("master"));
        assert!(!utils::is_likely_default_branch("feature-branch"));
        
        let default_branches = utils::get_common_default_branches();
        assert!(default_branches.contains(&"main".to_string()));
        assert!(default_branches.contains(&"master".to_string()));
        
        let dev_branches = utils::get_common_dev_branches();
        assert!(dev_branches.contains(&"develop".to_string()));
        assert!(dev_branches.contains(&"dev".to_string()));
        
        let comprehensive = utils::get_comprehensive_fallback_branches();
        assert!(comprehensive.len() > default_branches.len());
    }
}