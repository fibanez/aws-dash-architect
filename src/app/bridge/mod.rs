//! Bridge Module - AI Agent Tools for AWS Infrastructure Management
//!
//! This module provides AI agent tools that bridge natural language requests
//! with AWS resource operations through the Explorer system.

pub mod agents;
pub mod cancellation;
pub mod model_config;
pub mod performance;
pub mod sub_agent_callback_handler;
pub mod tools;
pub mod tools_registry;

pub use agents::*;
pub use cancellation::*;
pub use model_config::*;
pub use performance::*;
pub use sub_agent_callback_handler::*;
pub use tools::*;
pub use tools_registry::*;
