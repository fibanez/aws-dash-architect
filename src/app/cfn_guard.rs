//! CloudFormation Guard integration for policy-as-code validation.
//!
//! This module provides integration with AWS CloudFormation Guard for validating
//! CloudFormation templates against compliance rules and security policies.

use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::app::cfn_template::CloudFormationTemplate;

/// Main validator for CloudFormation Guard rules
#[derive(Debug, Clone)]
pub struct GuardValidator {
    /// Mapping of rule names to their content
    rules: HashMap<String, String>,
    /// Compliance programs enabled for validation
    compliance_programs: Vec<ComplianceProgram>,
}

/// Result of Guard validation containing violations and summary
#[derive(Debug, Clone)]
pub struct GuardValidation {
    /// List of violations found during validation
    pub violations: Vec<GuardViolation>,
    /// Whether the template is compliant (no violations)
    pub compliant: bool,
    /// Total number of rules evaluated
    pub total_rules: usize,
}

/// Individual violation found by Guard validation
#[derive(Debug, Clone)]
pub struct GuardViolation {
    /// Name of the rule that was violated
    pub rule_name: String,
    /// Name of the resource that violates the rule
    pub resource_name: String,
    /// Human-readable description of the violation
    pub message: String,
    /// Severity level of the violation
    pub severity: ViolationSeverity,
}

/// Severity levels for Guard violations
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ViolationSeverity {
    Critical,
    High,
    Medium,
    Low,
}

/// Available compliance programs for Guard validation
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum ComplianceProgram {
    NIST80053R4,
    NIST80053R5,
    NIST800171,
    PCIDSS,
    HIPAA,
    SOC,
    FedRAMP,
    Custom(String),
}

impl GuardValidator {
    /// Create a new Guard validator with the specified compliance programs
    pub async fn new(compliance_programs: Vec<ComplianceProgram>) -> Result<Self> {
        // For now, return a placeholder implementation
        // TODO: Load actual rules from compliance programs
        Ok(GuardValidator {
            rules: HashMap::new(),
            compliance_programs,
        })
    }

    /// Get the list of compliance programs enabled for this validator
    pub fn get_compliance_programs(&self) -> &[ComplianceProgram] {
        &self.compliance_programs
    }

    /// Validate a CloudFormation template against the loaded Guard rules
    pub async fn validate_template(&self, template: &CloudFormationTemplate) -> Result<GuardValidation> {
        // For now, return a placeholder implementation
        // TODO: Integrate with actual cfn-guard library
        
        // Convert template to JSON for Guard validation
        let _template_json = serde_json::to_string_pretty(template)?;
        
        // Placeholder: Check for common security issues
        let violations = self.check_placeholder_violations(template);
        let compliant = violations.is_empty();
        
        Ok(GuardValidation {
            violations,
            compliant,
            total_rules: self.rules.len(),
        })
    }

    /// Placeholder method for basic violation checking
    /// TODO: Replace with actual Guard validation
    fn check_placeholder_violations(&self, template: &CloudFormationTemplate) -> Vec<GuardViolation> {
        let mut violations = Vec::new();
        
        // Check for S3 buckets without encryption (placeholder rule)
        for (resource_name, resource) in &template.resources {
            if resource.resource_type == "AWS::S3::Bucket" {
                // Check if PublicReadPolicy is enabled (insecure)
                if let Some(public_read) = resource.properties.get("PublicReadPolicy") {
                    if public_read.as_bool() == Some(true) {
                        violations.push(GuardViolation {
                            rule_name: "S3_BUCKET_PUBLIC_READ_PROHIBITED".to_string(),
                            resource_name: resource_name.clone(),
                            message: "S3 bucket should not allow public read access".to_string(),
                            severity: ViolationSeverity::High,
                        });
                    }
                }
            }
        }
        
        violations
    }
}