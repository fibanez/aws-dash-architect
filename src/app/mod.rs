//! Core application modules for AWS Dash Architect.
//!
//! This module contains the business logic and data models that power the AWS CloudFormation
//! template designer and resource management system. Each submodule handles a specific aspect
//! of CloudFormation template processing, AWS integration, or user interface management.
//!
//! # Module Organization
//!
//! ## CloudFormation Processing
//! - [`cfn_template`] - Core CloudFormation template representation and manipulation
//! - [`cfn_resources`] - AWS resource specifications and schema management
//! - [`cfn_dag`] - Resource dependency graph analysis and validation
//! - [`cfn_intrinsic_functions`] - CloudFormation function detection and processing
//! - [`cfn_resource_icons`] - AWS resource type to icon mappings
//! - [`cfn_resource_policies`] - Resource policy management and validation
//!
//! ## AWS Integration
//! - [`aws_identity`] - AWS Identity Center authentication and credential management
//! - [`bedrock_client`] - AWS Bedrock integration for AI-powered features
//!
//! ## Project and Data Management
//! - [`projects`] - Multi-environment project organization and persistence
//! - [`dashui`] - Complete user interface implementation with window management
//!
//! # Integration Patterns
//!
//! The modules follow a layered architecture where:
//! - Data models ([`cfn_template`], [`projects`]) provide core representations
//! - Processing modules ([`cfn_dag`], [`cfn_resources`]) handle business logic
//! - Integration modules ([`aws_identity`], [`bedrock_client`]) connect to external services
//! - UI modules ([`dashui`]) coordinate user interactions and visual presentation
//!
//! See the [technical documentation](../../docs/technical/index.wiki) for detailed
//! implementation guides and architectural patterns for each subsystem.

pub mod aws_identity;
pub mod bedrock_client;
pub mod bridge;
pub mod cf_syntax;
pub mod cfn_dag;
pub mod cfn_intrinsic_functions;
pub mod cfn_resource_icons;
pub mod cfn_resource_policies;
pub mod cfn_resources;
pub mod cfn_template;
pub mod cloudformation_manager;
pub mod dashui;
pub mod notifications;
pub mod projects;
pub mod resource_explorer;

pub use dashui::app::DashApp;
