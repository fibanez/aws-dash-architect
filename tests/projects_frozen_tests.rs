//! Project Management System Frozen Tests
//!
//! This module provides comprehensive frozen testing for the project management
//! data structures that store user projects, environments, and CloudFormation
//! resources. These tests ensure that users' complex project configurations
//! remain stable and accessible across application updates.
//!
//! # Why Project Data Integrity Matters
//!
//! Users invest significant time organizing their CloudFormation infrastructure
//! into projects with multiple environments, resource configurations, and deployment
//! settings. Project data represents:
//! - Hours of manual configuration and organization work
//! - Critical business infrastructure mapped to specific environments
//! - Complex multi-region and multi-account deployment architectures
//! - Historical resource relationships and dependency mappings
//!
//! # Data Structures Under Test
//!
//! The project management system includes several interconnected data types:
//! - **Project**: Top-level container with metadata, environments, and templates
//! - **Environment**: AWS account and region configurations for deployment contexts
//! - **CloudFormationResource**: Individual AWS resources with properties and metadata
//! - **ResourceNode**: Dependency graph nodes representing resource relationships
//! - **AwsRegion/AwsAccount**: AWS service configuration primitives
//!
//! # Serialization Format Validation
//!
//! Project data is serialized to JSON for:
//! - **Persistent storage**: User projects saved to disk between application sessions
//! - **Data export**: Sharing project configurations between team members
//! - **Backup and restore**: Protecting user investment in project organization
//! - **Version control integration**: Tracking project changes over time
//!
//! Any changes to serialization format could result in:
//! - Loss of user project configurations requiring manual recreation
//! - Incompatibility with team-shared project files
//! - Broken integration with external tools consuming project data
//! - Corruption of resource relationship mappings
//!
//! # Testing Methodology
//!
//! These frozen tests use `insta` snapshot testing to capture exact JSON
//! representations of project data structures. Each test validates:
//! 1. Complete data structure serialization including all fields
//! 2. Proper handling of optional fields and nested relationships
//! 3. Compatibility with complex real-world project configurations
//! 4. Preservation of user-critical metadata and configuration details

use awsdash::app::cfn_template::CloudFormationTemplate;
use awsdash::app::projects::{
    AwsAccount, AwsRegion, CloudFormationResource, Environment, Project, ResourceNode,
};
use chrono::Utc;
use insta::assert_json_snapshot;
use std::collections::HashMap;
use std::path::PathBuf;

/// Verifies that complete Project data structure maintains serialization stability.
///
/// This test ensures that the main Project container preserves its format across
/// application updates, protecting users' comprehensive project configurations
/// from corruption. This matters because users organize complex infrastructure
/// projects with multiple environments, source control integration, and CloudFormation
/// templates that would be time-intensive to recreate.
///
/// # What This Test Covers
///
/// - **Project metadata**: Name, description, creation/update timestamps
/// - **Source integration**: Local folder paths and Git repository URLs
/// - **Environment configurations**: Development, staging, and production setups
/// - **Template integration**: Embedded CloudFormation templates (DAG now computed dynamically)
/// - **Regional defaults**: AWS region preferences for project deployments
///
/// # User Impact
///
/// If this test fails, users might experience:
/// - Loss of entire project configurations requiring complete rebuild
/// - Broken links to local folders and Git repositories
/// - Corruption of environment-specific deployment settings
/// - Invalid CloudFormation template associations
/// - Loss of project organization and metadata
///
/// # Complex Project Scenarios Tested
///
/// - Full project with all optional fields populated
/// - Multi-environment project structure (development environment shown)
/// - Integration with source control systems via Git URL
/// - Local filesystem integration via project folder paths
/// - Complete CloudFormation template embedding (DAG computed on-demand)
/// - Timestamp preservation for project lifecycle tracking
#[test]
fn test_project_structure() {
    let project = Project {
        name: "Example Project".to_string(),
        description: "A test project".to_string(),
        short_name: "example".to_string(),
        created: chrono::DateTime::parse_from_rfc3339("2025-01-01T10:00:00Z")
            .unwrap()
            .with_timezone(&Utc),
        updated: chrono::DateTime::parse_from_rfc3339("2025-01-01T10:00:00Z")
            .unwrap()
            .with_timezone(&Utc),
        local_folder: Some(PathBuf::from("/home/user/projects/example")),
        git_url: Some("https://github.com/user/example.git".to_string()),
        environments: vec![Environment {
            name: "Development".to_string(),
            aws_regions: vec![AwsRegion("us-east-1".to_string())],
            aws_accounts: vec![AwsAccount("123456789012".to_string())],
            deployment_status: None,
        }],
        default_region: Some("us-east-1".to_string()),
        cfn_template: Some(CloudFormationTemplate::default()),
    };

    // Test that project structure is frozen
    assert_json_snapshot!("project_structure", project);
}

/// Verifies that Environment configuration maintains stable serialization format.
///
/// This test ensures that Environment structures preserve their data layout,
/// protecting users' multi-environment deployment configurations from corruption.
/// This matters because users rely on environment-specific settings to deploy
/// the same CloudFormation templates across development, staging, and production
/// environments with different AWS accounts and regions.
///
/// # What This Test Covers
///
/// - **Multi-region deployment**: Support for deploying across multiple AWS regions
/// - **Cross-account access**: Configuration for different AWS accounts per environment
/// - **Environment naming**: Human-readable environment identification
/// - **Regional redundancy**: Multi-region disaster recovery and deployment strategies
///
/// # User Impact
///
/// If this test fails, users might experience:
/// - Loss of environment-specific deployment configurations
/// - Inability to deploy to multiple regions or accounts
/// - Corruption of staging and production environment settings
/// - Required manual reconfiguration of complex deployment pipelines
///
/// # Multi-Environment Deployment Scenarios
///
/// - Staging environment with multiple regions (us-west-2, eu-central-1)
/// - Cross-account deployment configuration for isolated environments
/// - Regional failover and disaster recovery configurations
/// - Environment-specific account and region combinations
#[test]
fn test_environment_structure() {
    let env = Environment {
        name: "Staging".to_string(),
        aws_regions: vec![
            AwsRegion("us-west-2".to_string()),
            AwsRegion("eu-central-1".to_string()),
        ],
        aws_accounts: vec![AwsAccount("111222333444".to_string())],
        deployment_status: None,
    };

    // Test that environment structure is frozen
    assert_json_snapshot!("environment_structure", env);
}

/// Verifies that CloudFormationResource maintains stable property serialization.
///
/// This test ensures that individual CloudFormation resources preserve their
/// structure and property formats, protecting users' detailed resource configurations
/// from corruption. This matters because users invest significant time configuring
/// resource properties, and any format changes could corrupt their infrastructure
/// definitions and prevent successful deployments.
///
/// # What This Test Covers
///
/// - **Resource identification**: Logical ID and AWS resource type preservation
/// - **Property serialization**: Complex property values and nested structures
/// - **Dynamic property handling**: Support for arbitrary property types and values
/// - **CloudFormation compatibility**: Ensures resource format matches AWS expectations
///
/// # User Impact
///
/// If this test fails, users might experience:
/// - Corruption of individual resource configurations
/// - Loss of carefully configured resource properties
/// - Deployment failures due to malformed resource definitions
/// - Incompatibility with AWS CloudFormation service expectations
///
/// # Resource Configuration Scenarios
///
/// - S3 Bucket resource with custom properties (BucketName configuration)
/// - Property handling for string values and complex nested objects
/// - Resource type validation for AWS service compatibility
/// - Logical ID preservation for CloudFormation template references
#[test]
fn test_cloudformation_resource() {
    let mut resource =
        CloudFormationResource::new("MyS3Bucket".to_string(), "AWS::S3::Bucket".to_string());
    resource.properties = HashMap::from([(
        "BucketName".to_string(),
        serde_json::Value::String("my-bucket".to_string()),
    )]);

    // Test that resource structure is frozen
    assert_json_snapshot!("cloudformation_resource", resource);
}

/// Verifies that ResourceNode dependency relationships maintain stable format.
///
/// This test ensures that dependency graph nodes preserve their structure,
/// protecting users' carefully mapped resource relationships from corruption.
/// This matters because users rely on accurate dependency tracking to understand
/// deployment order, troubleshoot failures, and optimize CloudFormation stack updates.
///
/// # What This Test Covers
///
/// - **Resource identification**: Unique resource ID preservation in dependency graphs
/// - **Dependency relationships**: Explicit depends_on relationships between resources
/// - **Graph node structure**: Basic building blocks for dependency analysis algorithms
/// - **Deployment ordering**: Critical data for CloudFormation deployment sequencing
///
/// # User Impact
///
/// If this test fails, users might experience:
/// - Corruption of resource dependency relationships
/// - Incorrect deployment order causing stack failures
/// - Loss of dependency analysis and visualization
/// - Broken resource relationship tracking for troubleshooting
///
/// # Dependency Management Scenarios
///
/// - EC2 Instance depending on Security Group (MyInstance -> MySecurityGroup)
/// - Explicit dependency declaration for deployment ordering
/// - Resource relationship preservation for graph algorithms
/// - Dependency chain validation for CloudFormation compatibility
#[test]
fn test_resource_node() {
    let node = ResourceNode {
        resource_id: "MyInstance".to_string(),
        depends_on: vec!["MySecurityGroup".to_string()],
    };

    // Test that node structure is frozen
    assert_json_snapshot!("resource_node", node);
}

/// Verifies that AwsRegion equality comparison works correctly for environment management.
///
/// This test ensures that AWS region comparison logic functions properly for
/// environment configuration and resource deployment targeting. This matters
/// because users depend on accurate region matching to deploy resources to
/// the correct AWS regions and avoid costly deployment mistakes.
///
/// # What This Test Covers
///
/// - **Region identity comparison**: Same region strings should be equal
/// - **Region distinction**: Different region strings should not be equal
/// - **Environment configuration**: Foundation for multi-region environment setup
/// - **Deployment targeting**: Critical for ensuring resources deploy to correct regions
///
/// # User Impact
///
/// If this test fails, users might experience:
/// - Incorrect region targeting in multi-region deployments
/// - Environment configuration errors with duplicate regions
/// - Deployment failures due to region mismatch logic
/// - Resource provisioning in wrong AWS regions
///
/// # Regional Configuration Scenarios
///
/// - Identical region comparison (us-east-1 == us-east-1)
/// - Different region distinction (us-east-1 != eu-west-1)
/// - Foundation for environment region validation
/// - Multi-region deployment logic correctness
#[test]
fn test_aws_region_equality() {
    let region1 = AwsRegion("us-east-1".to_string());
    let region2 = AwsRegion("us-east-1".to_string());
    let region3 = AwsRegion("eu-west-1".to_string());

    assert_eq!(region1, region2);
    assert_ne!(region1, region3);
}
