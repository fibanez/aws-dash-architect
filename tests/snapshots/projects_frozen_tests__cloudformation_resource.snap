---
source: tests/projects_frozen_tests.rs
expression: resource
---
{
  "resource_id": "MyS3Bucket",
  "resource_type": "AWS::S3::Bucket",
  "properties": {
    "BucketName": "my-bucket"
  }
}
